### ISSUE TITLE
Option to Roll Back Last Update

### ISSUE THREAD

**Issue by cmeans** [microsoft/vscode#77953]
Same as https://github.com/microsoft/vscode/issues/49442, but it's locked...so no-one has a way to bump up its priority.

Users need a way to revert the last update applied.

I applied the most recent update, and restarted the app...now I cannot do my work.  Will have to decide if a tool that breaks on me is really a good choice going forward.

If this isn't fixed soon, I will have no choice but to completely abandon this tool.

**Comment by pbysh** [comment]
I agree, I LOVE VS Code and specifically the Remote-SSH and WSL tools, they have made my development experience on Windows amazing. But today my Remote-SSH is broken (same as you, I believe) and now I'm kinda sorta stuck because I can't easily revert back to yesterday's build.

**Comment by forrcaho** [comment]
FWIW, The regular release of VS Code now supports the Remote Development extensions. I also had Remote-SSH break with the Insiders update today (where the hell do you even report that?) and went back to the standard release (which I'll probably stick with now, unless the rollback feature mentioned in _this_ ticket is implemented).

**Comment by beastdestroyer** [comment]
Any bug report, for stable, insiders or exploration you report in here, the vscode repo; or if you believe it is extension-based, then report it in its repo

**Comment by joaomoreno** [comment]
> I applied the most recent update, and restarted the app...now I cannot do my work.

@cmeans What issue did you hit? Are you on Insiders?

> But today my Remote-SSH is broken (same as you, I believe) and now I'm kinda sorta stuck because I can't easily revert back to yesterday's build.

@pbysh What issue is that?

---

The policy is: we want people moving forward with us all the time. We're all Insiders selfhosters in the team. If something breaks severely, it will be fixed in no time. If something breaks mildly, it will be fixed eventually.

Insiders is an aggressive program and so it is optional, you don't need to stick to it, although we do appreciate if you do. When things break badly, there's always VS Code stable as well as ZIP downloads which do not automatically update.

**Comment by forrcaho** [comment]
After yesterday's (2019-07-25) Insider's upgrade, when I open VSCode, which is set to connect to a remote machine via Remote-SSH, I get an info message in the lower right corner which says "Setting up SSH Host sandbox: (details) Initializing VS Code Server". It hangs there. The blue lines sweeping below the info box and above the editor window continue indefinitely.

The info message details link shows the following output in the Terminal window:
```
PWD=/home/forrest
LANG=en_US.UTF-8
HOME=/home/forrest
SHLVL=2
LOGNAME=forrest
TMP=/tmp/user/2239
SSH_CONNECTION=10.12.100.6 59126 10.8.134.62 22
XDG_RUNTIME_DIR=/run/user/2239
_=/usr/bin/printenv
Starting server...
Waiting for server log...
 
*
* Reminder: You may only use this software with Visual Studio family products,
* as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)
*
 
ba58e60a-0522-4d69-b87f-9931c9a7d0a6==43635==46184==
```
That's all, for as long as I care to wait.

I killed the remote vscode-insiders server, but got the same result.

I observed this behavior on a Mac OS X host connecting to a remote VM running Ubuntu, and from a Debian "sid" host connecting to another remote Debian "sid" computer (not a VM).

**Comment by forrcaho** [comment]
Today's Insiders release (commit 6d82826e990407364d5c829b73602a19ecdf7d00) fixes the Remote - SSH breakage from yesterday's release (commit d4b3a7a22fe01ce6af85f9c59943ad54d6f6f205). That's great! However, some of us are using Insiders to do actual work and it would be infinitely helpful to be able to roll back to the previously installed version so we're not suddenly SOL in our development work. Having such a feature could easily mean the difference between having people really use Insiders and thus give it a real testing, and only having people who don't have real work to do test it. As a bonus, you could add telemetry and see when a release is being rolled back _en masse_, and what extensions those rolling back have installed.

**Comment by wezleytsai** [comment]
I would argue that an easy rollback mechanism actually encourages people to keep up with the latest version, as it lowers the risk of trying an update. Let me share my experience:

I was on 1.37.1 and I've always had automatic updates on which was never a problem. I then upgraded to 1.38 at one point, and one of my frequent keyboard shortcuts stopped working. I do ctrl+D to peek definition, and the subsequently enter key to go to that definition if I so choose. This stopped working, peek definition worked, but enter key did nothing. I spent a little bit of time looking into the issue, hacked around my keyboard shortcuts a bit to get it to "work" until I realized my fix broke other behaviors. I even created an [issue](https://github.com/microsoft/vscode/issues/80822), but it got closed as duplicate without any reference to the other ticket. I had no issue to subscribe to to get notified of a fix. I eventually gave up and manually downgraded to 1.37.1 and turned off automatic updates. Only today did I decide to turn on updates, upgrade to 1.40, and found that the issue has been resolved. I enjoy VS Code immensely and would not consider it abandoning it over bugs like this, _but_ if there was an easier way to rollback, I probably would have upgraded to 1.38.x, 1.39 and so forth instead of skipping some probably great updates.
